pkg/distributor/otel.go

```diff
@@ -252,12 +252,10 @@ func otlpHandler(
 				grpcCode codes.Code
 				errorMsg string
 			)
-			if resp, ok := httpgrpc.HTTPResponseFromError(err); ok {
-				// here the error would always be nil, since it is already checked in httpgrpc.HTTPResponseFromError
-				s, _ := grpcutil.ErrorToStatus(err)
-				httpCode = int(resp.Code)
-				grpcCode = s.Code() // this will be the same as httpCode.
-				errorMsg = string(resp.Body)
+			if st, ok := grpcutil.ErrorToStatus(err); ok {
+				httpCode = int(st.Code())
```

Comment:
Confirmed with @ying-jeanne, I think a problem here (carried over from @ying-jeanne's previous PR) is that `httpCode` doesn't get translated with regards to OTel's retriable HTTP status codes.

Should we possibly instead just map retryable Prometheus HTTP status codes to OTel ones?

Reply:
@ying-jeanne was the goal of https://github.com/grafana/mimir/pull/8324 to map **all** `5xx` HTTP status codes that are non-retryable according to the [OTLP specification](https://opentelemetry.io/docs/specs/otlp/#failures-1) to retryable ones?
So far, the actual delta of https://github.com/grafana/mimir/pull/8324 is that `500` is mapped to `503`. But all other `5xx` errors different from `502`, `503` and `504` are still non-retryable (for example `501`). What do we want to do with `501`? Do we want it to be retryable or non-retryable?

Reply:
After a discussion in Slack we agreed that the mapping from Prometheus retryable HTTP status codes to OTLP retryable HTTP status codes should be:
| Prometheus HTTP status code  | OTLP HTTP status code  |
|--------------------------------|--------------------------------|
| `2xx` | `2xx` |
| `4xx` | `4xx` |
| `500` | `503` |
| `501` | `503` |
| `502` | `502` |
| `503` | `503` |
| `504` | `504` |
| >`504` | `503` |

Reply:
> @ying-jeanne was the goal of #8324 to map **all** `5xx` HTTP status codes that are non-retryable according to the [OTLP specification](https://opentelemetry.io/docs/specs/otlp/#failures-1) to retryable ones? So far, the actual delta of #8324 is that `500` is mapped to `503`. But all other `5xx` errors different from `502`, `503` and `504` are still non-retryable (for example `501`). What do we want to do with `501`? Do we want it to be retryable or non-retryable?

The original function contains only 500 and 501, and since 501 is `not implemented` and align with loki's fix https://github.com/grafana/loki/pull/13173/files, we just map 500 to 503. But now the fix is on all path, we should at least map 504+ to 503, and 429 to 529

Reply:
The [table](https://github.com/grafana/mimir/pull/8339#discussion_r1636246402) I posted above shows the current mimir mapping. We map 501 (coming from ingesters) to 503 (OTLP) because Prometheus retries 501 errors too.